src/proper_statem.erl

```diff
@@ -463,6 +493,188 @@ move_shrinker(Seq, Par, I) ->
 	 move_shrinker(NewSeq, NewPar, I-1)).
 
 
+%% -----------------------------------------------------------------------------
+%% Targeted command generation
+%% -----------------------------------------------------------------------------
+
+
+%% @doc A special PropEr type which generates random command sequences,
+%% according to an abstract state machine specification and taking into
+%% consideration a utility value. The function takes as input the name of
+%% a callback module, which contains the state machine specification. The
+%% initial state is computed by `Mod:initial_state/0'.
+
+-spec targeted_commands(mod_name()) -> proper_types:type().
+targeted_commands(Mod) ->
+  proper_types:add_prop(is_user_nf_stateful, true,
+                        ?USERNF(targeted_commands_gen(Mod),
+                                next_commands_gen(Mod))).
+
+%% @doc Similar to {@link targeted_commands/1}, but generated command sequences
+%% always start at a given state. In this case, the first command is always
+%% `{init, InitialState}' and is used to correctly initialize the state
+%% every time the command sequence is run (i.e. during normal execution,
+%% while shrinking and when checking a counterexample). In this case,
+%% `Mod:initial_state/0' is not called.
+
+-spec targeted_commands(mod_name(), symbolic_state()) -> proper_types:type().
+targeted_commands(Mod, InitialState) ->
+  proper_types:add_prop(is_user_nf_stateful, true,
+                        ?USERNF(targeted_commands_gen(Mod, InitialState),
+                                next_commands_gen(Mod, InitialState))).
+
+%% @private
+-spec targeted_commands_gen(mod_name()) -> proper_types:type().
+targeted_commands_gen(Mod) ->
+  InitialState = Mod:initial_state(),
+  ?LET(Cmds, ?LAZY(commands(Mod)),
+       {finalize_weights(Mod, InitialState, Cmds, maps:new()), Cmds}).
+
+%% @private
+-spec targeted_commands_gen(mod_name(), symbolic_state()) ->
+        proper_types:type().
+targeted_commands_gen(Mod, InitialState) ->
+  ?LET(Cmds, ?LAZY(commands(Mod, InitialState)),
+       {finalize_weights(Mod, InitialState, Cmds, maps:new()), Cmds}).
+
+%% @private
+-spec next_commands_gen(mod_name()) -> next_fun().
+next_commands_gen(Mod) ->
```

Comment:
Isn't this better written as:
```erlang
next_commands_gen(Mod) ->
  next_commands_gen(Mod, Mod:initial_state()).
```
or is there some other difference in the two functions that I am missing?

Reply:
There is a difference (similar to what already exists for the simple `commands/1` and `commands/2`). If not providing an initial state the commands returned are:

```erlang
[{set, {var,1}, {call,mod,some_call,[some_args]}}, ...]
```

If an initial state is provided then the commands returned are:

```erlang
[{init,some_initial_state}, {set, {var,1}, {call,mod,some_call,[some_args]}}, ...]
```